# Welcome to Advent of Code for the Year 2025

## aoc-1.1
The Elves have good news and bad news.

The good news is that they've discovered project management! This has given them the tools they need to prevent their usual Christmas emergency. For example, they now know that the North Pole decorations need to be finished soon so that other critical tasks can start on time.

The bad news is that they've realized they have a different emergency: according to their resource planning, none of them have any time left to decorate the North Pole!

To save Christmas, the Elves need you to finish decorating the North Pole by December 12th.

Collect stars by solving puzzles. Two puzzles will be made available on each day; the second puzzle is unlocked when you complete the first. Each puzzle grants one star. Good luck!

You arrive at the secret entrance to the North Pole base ready to start decorating. Unfortunately, the password seems to have been changed, so you can't get in. A document taped to the wall helpfully explains:

"Due to new security protocols, the password is locked in the safe below. Please see the attached document for the new combination."

The safe has a dial with only an arrow on it; around the dial are the numbers 0 through 99 in order. As you turn the dial, it makes a small click noise as it reaches each number.

The attached document (your puzzle input) contains a sequence of rotations, one per line, which tell you how to open the safe. A rotation starts with an L or R which indicates whether the rotation should be to the left (toward lower numbers) or to the right (toward higher numbers). Then, the rotation has a distance value which indicates how many clicks the dial should be rotated in that direction.

So, if the dial were pointing at 11, a rotation of R8 would cause the dial to point at 19. After that, a rotation of L19 would cause it to point at 0.

Because the dial is a circle, turning the dial left from 0 one click makes it point at 99. Similarly, turning the dial right from 99 one click makes it point at 0.

So, if the dial were pointing at 5, a rotation of L10 would cause it to point at 95. After that, a rotation of R5 could cause it to point at 0.

The dial starts by pointing at 50.

You could follow the instructions, but your recent required official North Pole secret entrance security training seminar taught you that the safe is actually a decoy. The actual password is the number of times the dial is left pointing at 0 after any rotation in the sequence.

For example, suppose the attached document contained the following rotations:

L68
L30
R48
L5
R60
L55
L1
L99
R14
L82
Following these rotations would cause the dial to move as follows:

The dial starts by pointing at 50.
The dial is rotated L68 to point at 82.
The dial is rotated L30 to point at 52.
The dial is rotated R48 to point at 0.
The dial is rotated L5 to point at 95.
The dial is rotated R60 to point at 55.
The dial is rotated L55 to point at 0.
The dial is rotated L1 to point at 99.
The dial is rotated L99 to point at 0.
The dial is rotated R14 to point at 14.
The dial is rotated L82 to point at 32.
Because the dial points at 0 a total of three times during this process, the password in this example is 3.

Analyze the rotations in your attached document. What's the actual password to open the door?

### Input

In [2]:
path = '/mnt/c/Users/guspa/Desktop/python_homework/data/aoc25/aoc25_1.1.txt'
mode = 'r'

move_side = []
move_num = []

with open(path, mode) as file:
    for line in file.readlines():
        line = line.strip()

        move_side.append(line[0])

        move_num.append(line[1:])

print(move_side)
print(move_num)

['L', 'L', 'L', 'L', 'L', 'L', 'L', 'L', 'L', 'R', 'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'L', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'R', 'L', 'L', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'L', 'L', 'R', 'R', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'L', 'R', 'L', 'R', 'R', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'R', 'R', 'R', 'L', 'L', 'L', 'L', 'L', 'L', 'R', 'L', 'R', 'L', 'L', 'R', 'L', 'R', 'R', 'R', 'L', 'R', 'L', 'R', 'L', 'L', 'R', 'L', 'L', 'R', 'R', 'L', 'R', 'R', 'L', 'L', 'L', 'L', 'R', 'L', 'R', 'R', 'L', 'L', 'L', 'L', 'L', 'R', 'L', 'L', 'R', 'L', 'R', 'L', 'R', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'L', 'R', 'L', 'R', 'R', 'R', 'L', 'R', 'R', 'R', 'L', 'R', 'L', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'L', 'L', 'L', 'R', 'L', 'L', 'R',

### Functions

In [182]:
# closing a loop 0-99-0
def ouroboros(integer):
    if integer >= 100:
            integer = integer - 100
        
    elif integer < 0:
        integer = integer + 100

    else:
        integer = integer

    return integer

print('test:', ouroboros(123))

test: 23


In [183]:
# accounting for higher number than 99
def extreme_moves(integer):
    if integer >= 100:
        integer = integer % 100
    else:
        integer = integer
    
    return integer

print('test:', extreme_moves(456))

test: 56


### Programme

In [184]:
password = 0
position = 50

for turn in range(len(move_num)):

    move = extreme_moves(int(move_num[turn]))

    if move_side[turn] == 'L':
        position = position - move

        position = ouroboros(integer = position)

        print(position)

    else:
        position = position + move

        position = ouroboros(integer = position)
        
        print(position)

    if position == 0:
        password += 1

30
17
1
85
56
49
1
53
22
49
25
21
87
46
36
42
85
14
39
43
30
37
46
4
46
30
11
0
92
95
97
14
23
64
6
92
88
10
91
42
13
36
56
43
25
19
79
64
9
43
98
61
52
23
0
42
14
17
0
79
72
6
47
90
41
92
0
57
0
54
0
49
80
24
29
0
80
5
25
67
0
74
62
32
69
0
75
58
0
36
84
17
61
0
11
44
53
0
66
5
0
92
42
50
0
65
0
59
49
85
41
98
0
57
58
0
38
0
4
0
21
0
65
0
1
0
34
52
8
0
31
23
44
0
35
27
0
52
12
23
58
0
77
45
21
70
0
74
6
35
15
84
0
7
97
71
0
87
0
79
15
0
94
0
55
42
73
63
75
36
28
87
0
86
36
0
23
56
60
48
60
83
70
0
74
0
77
78
0
25
81
1
20
0
22
0
1
0
19
69
20
16
77
18
56
27
68
26
51
0
33
50
58
67
0
19
98
77
36
19
36
90
32
0
3
65
88
0
43
0
96
58
0
21
0
14
0
95
73
4
55
0
34
0
16
0
52
3
14
12
37
76
46
56
1
28
10
35
75
0
51
85
41
97
76
0
11
0
47
0
40
0
55
63
0
89
69
72
0
12
96
41
6
45
90
62
41
0
48
78
80
12
80
4
53
41
36
66
0
44
80
22
0
53
78
12
5
0
63
9
62
96
69
64
21
26
0
38
0
11
66
11
0
12
25
50
79
95
22
54
27
18
0
57
0
63
60
0
71
4
7
2
3
49
63
0
78
85
69
58
0
38
76
6
78
55
79
13
0
30
0
83
39
0
38
76
57


### Evaluation of Results

In [185]:
if password == 0:
    print('There is no decoded password.')
else:
    print('Password:', password)

Password: 1182


### Deconstruction of the Example

In [186]:
# we need:
# move  | position
# 0     | 50
# L68   | 82
# L30   | 52
# R48   | 0
# L5    | 95
# R60   | 55
# L55   | 0
# L1    | 99
# L99   | 0
# R14   | 14
# L82   | 32
# R123  | 55

# password : 3

# what we have:
# input sequence of moves as strings in a list - numbers
move_num = ['68', '30', '48', '5', '60', '55', '1', '99', '14', '82', '123']
# input sequence of moves as strings in a list - sides
move_side = ['L', 'L', 'R', 'L', 'R', 'L', 'L', 'L', 'R', 'L', 'R']
# accumulator variable for password
password = 0
# starting position
position = 50

# for each turn in range of length of move_num sequence (11), do
for turn in range(len(move_num)):

    # if integer at index position (0) of move_num sequence (68) is larger or equal to 100, do
    if int(move_num[turn]) >= 100:
        # move becomes remainig of integer at index position of move_num sequence divided by 100
        move = int(move_num[turn]) % 100
    # otherwise, do
    else:
        # move becomes remainig of integer at index position of move_num sequence (68)
        move = int(move_num[turn])

    # if value at index position (0) of move_side sequence (L) is L, do
    if move_side[turn] == 'L':
        # position becomes a result of subtraction of move (68) from position (50)
        position = position - move
        # if position is larger or equal to 100, do
        if position >= 100:
            # position becomes a result of subtraction of 100 from position
            position = position - 100
        # else if position is smaller than 0, do
        elif position < 0:
            # position becomes a result of addition of 100 to position (-18)
            position = position + 100
        # otherwise, do
        else:
            #position becomes position
            position = position
        # debugging
        print(position)
    # otherwise, do
    else:
        # position becomes a result of addition of move to position
        position = position + move
        # if position is larger or equal to 100, do
        if position >= 100:
            # position becomes a result of subtraction of 100 from position
            position = position - 100
        # else if position is smaller than 0, do
        elif position < 0:
            # position becomes a result of addition of 100 to position (-18)
            position = position + 100
        # otherwise, do
        else:
            #position becomes position
            position = position
        # debugging
        print(position)

    # if position is 0, do
    if position == 0:
        # add 1 to password
        password += 1

# if password is 0, do
if password == 0:
    # print option with no password
    print('There is no decoded password.')
# otherwise, do
else:
    # print option with password
    print('Password:', password)

82
52
0
95
55
0
99
0
14
32
55
Password: 3


### Answer
1182

## aoc25-1.2
You're sure that's the right password, but the door won't open. You knock, but nobody answers. You build a snowman while you think.

As you're rolling the snowballs for your snowman, you find another security document that must have fallen into the snow:

"Due to newer security protocols, please use password method 0x434C49434B until further notice."

You remember from the training seminar that "method 0x434C49434B" means you're actually supposed to count the number of times any click causes the dial to point at 0, regardless of whether it happens during a rotation or at the end of one.

Following the same rotations as in the above example, the dial points at zero a few extra times during its rotations:

The dial starts by pointing at 50.
The dial is rotated L68 to point at 82; during this rotation, it points at 0 once.
The dial is rotated L30 to point at 52.
The dial is rotated R48 to point at 0.
The dial is rotated L5 to point at 95.
The dial is rotated R60 to point at 55; during this rotation, it points at 0 once.
The dial is rotated L55 to point at 0.
The dial is rotated L1 to point at 99.
The dial is rotated L99 to point at 0.
The dial is rotated R14 to point at 14.
The dial is rotated L82 to point at 32; during this rotation, it points at 0 once.
In this example, the dial points at 0 three times at the end of a rotation, plus three more times during a rotation. So, in this example, the new password would be 6.

Be careful: if the dial were pointing at 50, a single rotation like R1000 would cause the dial to point at 0 ten times before returning back to 50!

Using password method 0x434C49434B, what is the password to open the door?

### Input

In [ ]:
path = '/mnt/c/Users/guspa/Desktop/python_homework/data/aoc25/aoc25_1.1.txt'
mode = 'r'

move_side = []
move_num = []

with open(path, mode) as file:
    for line in file.readlines():
        line = line.strip()

        move_side.append(line[0])

        move_num.append(line[1:])

print(move_side)
print(move_num)

['L', 'L', 'L', 'L', 'L', 'L', 'L', 'L', 'L', 'R', 'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'L', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'R', 'L', 'L', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'L', 'L', 'R', 'R', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'L', 'R', 'L', 'R', 'R', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'R', 'R', 'R', 'L', 'L', 'L', 'L', 'L', 'L', 'R', 'L', 'R', 'L', 'L', 'R', 'L', 'R', 'R', 'R', 'L', 'R', 'L', 'R', 'L', 'L', 'R', 'L', 'L', 'R', 'R', 'L', 'R', 'R', 'L', 'L', 'L', 'L', 'R', 'L', 'R', 'R', 'L', 'L', 'L', 'L', 'L', 'R', 'L', 'L', 'R', 'L', 'R', 'L', 'R', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'L', 'R', 'L', 'R', 'R', 'R', 'L', 'R', 'R', 'R', 'L', 'R', 'L', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'L', 'L', 'L', 'R', 'L', 'L', 'R',

### Programme

In [ ]:
password = 0
position = 50

for turn in range(len(move_num)):

    # IS STARTING AT 0?
    if position == 0:

        if int(move_num[turn]) >= 100:

            passing_zero = int(move_num[turn]) // 100
 
            password = password + passing_zero
            passing_zero = 0

            move = int(move_num[turn]) % 100
            
        # MOVE IS WITHIN 0-99
        else:
            move = int(move_num[turn])
        
        # IS MOVE L?
        if move_side[turn] == 'L':
            position = position - move

            if position < 0:
                position = position + 100
                
        # MOVE IS R
        else:
            position = position + move

    # STARTING AT 1-99
    else:

        # IS MOVE EXTREME >= 100?
        if int(move_num[turn]) >= 100:

            passing_zero = int(move_num[turn]) // 100

            password = password + passing_zero
            passing_zero = 0

            move = int(move_num[turn]) % 100
        
        # MOVE IS WITHIN 0-99
        else:
            move = int(move_num[turn])
        
        # IS MOVE L?
        if move_side[turn] == 'L':
            position = position - move

            # IS NEGATIVE?
            if position < 0:
                position = position + 100
                password += 1

        # MOVE IS R
        else:
            position = position + move

            # IS LARGER THAN 100?
            if position > 100:
                position = position - 100
                password += 1

            # IS EQUAL TO 100?
            elif position == 100:
                position = position - 100

    # IS FINAL POSITION 0?
    if position == 0:
        password += 1

### Evaluation of Results

In [ ]:
print('Password:', password)

Password: 6907


Password: 6907


### Deconstrucion of the Example

In [ ]:
# we need:
# move  | position
# 0     | 50
# L68   | 82 - passed 0
# L30   | 52
# R48   | 0 - is 0
# L5    | 95
# R60   | 55 - passed 0
# L55   | 0 - is 0
# L1    | 99
# L99   | 0 - is 0
# R14   | 14 
# L82   | 32 - passed 0

# password : 6

# what we have:
#move_num = ['68', '30', '48', '5', '60', '55', '1', '99', '14', '82']
#move_side = ['L', 'L', 'R', 'L', 'R', 'L', 'L', 'L', 'R', 'L']

# special cases:
#move_num = ['50', '200', '211', '99']
#move_side = ['L', 'L', 'L', 'L']

# move  | position | password
#       | 50
# L50   | 0        | 1 - end
# L200  | 0        | 3 - passed 0 1x + end
# L211  | 89       | 5 - passed 0 2x
# L99   | 90       | 6- passed 0 1

# accumulator variable for the sums of 0 passed or reached
password = 0
# starting position
position = 50

# for each turn in range of all moves
for turn in range(len(move_num)):

    # IS STARTING AT 0?
    if position == 0:
        # IS MOVE EXTREME >= 100?
        if int(move_num[turn]) >= 100:

            passing_zero = int(move_num[turn]) // 100 # count number of whole circles

            password = password + passing_zero # add counts to password
            passing_zero = 0 # reset counter

            move = int(move_num[turn]) % 100 # use the raminder from division by 100 in another conditions
            
        # MOVE IS WITHIN 0-99
        else:
            move = int(move_num[turn]) # keep the move and use it in another conditions
        
        # IS MOVE L?
        if move_side[turn] == 'L':
            position = position - move

            if position < 0:
                position = position + 100
                

        # MOVE IS R
        else:
            position = position + move

    # STARTING AT 1-99
    else:
        # IS MOVE EXTREME >= 100?
        if int(move_num[turn]) >= 100:

            passing_zero = int(move_num[turn]) // 100 # count number of whole circles

            password = password + passing_zero # add counts to password
            passing_zero = 0 # reset counter

            move = int(move_num[turn]) % 100 # use the raminder from division by 100 in another conditions
        
        # MOVE IS WITHIN 0-99
        else:
            move = int(move_num[turn]) # keep the move and use it in another conditions
        
        # IS MOVE L?
        if move_side[turn] == 'L':
            position = position - move # get new position by subtraction 

            # IS NEGATIVE?
            if position < 0:
                position = position + 100 # add 100 to the position to close the loop
                password += 1 # add 1 to password

        # MOVE IS R
        else:
            position = position + move # get new position by addition

            # IS LARGER THAN 100?
            if position > 100:
                position = position - 100 # subtract 100 from the position to close the loop
                password += 1 # add 1 to password

            # IS EQUAL TO 100?
            elif position == 100:
                position = position - 100 # subtract 100 from the position to close the loop

    # IS FINAL POSITION 0?
    if position == 0:
        password += 1 # add 1 to password

# PASSWORD
print('Password:', password)

Password: 6907


### Answer
6907

## Version 2

 ### Input

In [ ]:
path = '/mnt/c/Users/guspa/Desktop/python_homework/data/aoc25/aoc25_1.2.txt'
mode = 'r'

move_side = []
move_num = []

with open(path, mode) as file:
    for line in file.readlines():
        line = line.strip()

        move_side.append(line[0])

        move_num.append(line[1:])

print(move_side)
print(move_num)

['L', 'L', 'L', 'L', 'L', 'L', 'L', 'L', 'L', 'R', 'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'L', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'R', 'L', 'L', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'L', 'L', 'R', 'R', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'L', 'R', 'L', 'R', 'R', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'L', 'R', 'R', 'L', 'L', 'R', 'R', 'R', 'L', 'L', 'L', 'L', 'L', 'L', 'R', 'L', 'R', 'L', 'L', 'R', 'L', 'R', 'R', 'R', 'L', 'R', 'L', 'R', 'L', 'L', 'R', 'L', 'L', 'R', 'R', 'L', 'R', 'R', 'L', 'L', 'L', 'L', 'R', 'L', 'R', 'R', 'L', 'L', 'L', 'L', 'L', 'R', 'L', 'L', 'R', 'L', 'R', 'L', 'R', 'L', 'L', 'L', 'R', 'R', 'R', 'R', 'R', 'L', 'R', 'L', 'R', 'R', 'R', 'L', 'R', 'R', 'R', 'L', 'R', 'L', 'L', 'R', 'L', 'L', 'L', 'L', 'R', 'L', 'L', 'L', 'R', 'L', 'L', 'R',

### Programme

In [ ]:
password = 0
position = 50

for turn in range(len(move_num)):

    if move_side[turn] == 'L':
        for click in range(int(move_num[turn])):
            position -= 1

            if position == -1:
                position = 99

            if position == 0:
                password += 1

    else:
        for click in range(int(move_num[turn])):
            position += 1
     
            if position == 100:
                position = 0

            if position == 0:
                password += 1

### Evaluation of Results

In [ ]:
if password == 0:
    print('There is no decoded password.')
else:
    print('Password:', password) 

Password: 6907


### Deconstruction of the Example

In [ ]:
# accumulator variable for password
password = 0
# starting position
position = 50

# example
move_num = ['68', '30', '48', '5', '60', '55', '1', '99', '14', '82']
move_side = ['L', 'L', 'R', 'L', 'R', 'L', 'L', 'L', 'R', 'L']

# test
#move_num = ['250', '99', '1', '1', '1']
#move_side = ['R', 'R', 'R', 'R', 'L']

# for each turn in range of all moves, do
for turn in range(len(move_num)):

    # if it is L move, do
    if move_side[turn] == 'L':
        # for each click in range of move number, do
        for click in range(int(move_num[turn])):
            # subtract 1 from position
            position -= 1

            # if position is -1, do
            if position == -1:
                position = 99 # position becomes 99

            # if position is 0, do 
            if position == 0:
                password += 1 # add 1 to password

    # it is R move, do
    else:
        # for each click in range of move number, do
        for click in range(int(move_num[turn])):
            # add 1 to position
            position += 1

            # if position is 100, do           
            if position == 100:
                position = 0 # position becomes 0

            # if position is 0, do
            if position == 0:
                password += 1 # add 1 to password

# print password
print('password:', password)

password: 6


### Answer
6907

## aoc-2.1
You get inside and take the elevator to its only other stop: the gift shop. "Thank you for visiting the North Pole!" gleefully exclaims a nearby sign. You aren't sure who is even allowed to visit the North Pole, but you know you can access the lobby through here, and from there you can access the rest of the North Pole base.

As you make your way through the surprisingly extensive selection, one of the clerks recognizes you and asks for your help.

As it turns out, one of the younger Elves was playing on a gift shop computer and managed to add a whole bunch of invalid product IDs to their gift shop database! Surely, it would be no trouble for you to identify the invalid product IDs for them, right?

They've even checked most of the product ID ranges already; they only have a few product ID ranges (your puzzle input) that you'll need to check. For example:

11-22,95-115,998-1012,1188511880-1188511890,222220-222224,
1698522-1698528,446443-446449,38593856-38593862,565653-565659,
824824821-824824827,2121212118-2121212124
(The ID ranges are wrapped here for legibility; in your input, they appear on a single long line.)

The ranges are separated by commas (,); each range gives its first ID and last ID separated by a dash (-).

Since the young Elf was just doing silly patterns, you can find the invalid IDs by looking for any ID which is made only of some sequence of digits repeated twice. So, 55 (5 twice), 6464 (64 twice), and 123123 (123 twice) would all be invalid IDs.

None of the numbers have leading zeroes; 0101 isn't an ID at all. (101 is a valid ID that you would ignore.)

Your job is to find all of the invalid IDs that appear in the given ranges. In the above example:

11-22 has two invalid IDs, 11 and 22.
95-115 has one invalid ID, 99.
998-1012 has one invalid ID, 1010.
1188511880-1188511890 has one invalid ID, 1188511885.
222220-222224 has one invalid ID, 222222.
1698522-1698528 contains no invalid IDs.
446443-446449 has one invalid ID, 446446.
38593856-38593862 has one invalid ID, 38593859.
The rest of the ranges contain no invalid IDs.
Adding up all the invalid IDs in this example produces 1227775554.

What do you get if you add up all of the invalid IDs?

### Input

In [ ]:
path = '/mnt/c/Users/guspa/Desktop/python_homework/data/aoc25/aoc25_2.1.txt'
mode = 'r'

database = []
item = ''

with open(path, mode) as file:
    for line in file.readlines():
        line = line.strip()
        
        for position in range(len(line)):

            if line[position] == ',':
                database.append(item)
                item = ''

            elif position == len(line)-1:
                item += line[position]
                database.append(item)
                item = ''

            else:
                item += line[position]

print(database)

['19391-47353', '9354357-9434558', '4646427538-4646497433', '273-830', '612658-674925', '6639011-6699773', '4426384-4463095', '527495356-527575097', '22323258-22422396', '412175-431622', '492524-611114', '77-122', '992964846-993029776', '165081-338962', '925961-994113', '7967153617-7967231799', '71518058-71542434', '64164836-64292066', '4495586-4655083', '2-17', '432139-454960', '4645-14066', '6073872-6232058', '9999984021-10000017929', '704216-909374', '48425929-48543963', '52767-94156', '26-76', '1252-3919', '123-228']


### Functions

In [ ]:
def get_range(s):
    
    s1 = ''
    s2 = ''
    dash = ''

    for character in range(len(s)):
        if dash == '':
            #print('con1')
            
            if s[character] == '-':
                #print('con1a')
                dash += '-'

            else:
                #print('con1b')
                s1 += s[character]
        
        else:
            #print('con2')
            s2 += s[character]

        if character == len(s)-1:
            #print('clean')
            dash = ''

    return int(s1), int(s2)+1

get_range('11-22') # a, b+1

(11, 23)

In [ ]:
def invalid_mirror(integer):
    
    pattern1 = ''
    pattern2 = ''

    if len(str(integer)) %2 == 0:
        half = len(str(integer)) // 2
        
        for position in range(half): # 0, 1, 2
            pattern1 += str(integer)[position]

        for position in range(half,len(str(integer))): # 3, 4, 5
            pattern2 += str(integer)[position]

        if pattern1 == pattern2:

            return integer

        else: 
            return False
    else:
        return False

invalid_mirror(432432)

432432

### Programme

In [ ]:
invalid_ID = 0

for item in database:

    for value in range(get_range(item)[0], get_range(item)[1]):

        if value == invalid_mirror(value):
            invalid_ID += value
        else:
            continue

KeyboardInterrupt: 

### Evaluation of Results

In [ ]:
print('sum of invalid IDs:', invalid_ID)

sum of invalid IDs: 24747430309


### Deconstruction of the Example

In [ ]:
# database of examaple ranges
database = ['11-22', '95-115', '998-1012', '1188511880-1188511890', '222220-222224', '1698522-1698528', '446443-446449', '38593856-38593862', '565653-565659', '824824821-824824827', '2121212118-2121212124']

# accumulator variable for sums of invalid ID numbers
invalid_ID = 0

# for each item in databse, do
for item in database:
    # for each value within a range which starts with first part of the ID and ends with the second part of the ID including
    for value in range(get_range(item)[0], get_range(item)[1]):
        # if value is equal to value returned from a function which checks if the value could be invalid, do
        if value == invalid_mirror(value):
            print(value)
            invalid_ID += value # add the value to the accumualator variable
        # is not equal, do
        else:
            continue # go into another iteration (new number within the interval)
    
# print the sum of invalid IDs
print('sum of invalid IDs:', invalid_ID)

11
22
99
1010
1188511885
222222
446446
38593859
sum of invalid IDs: 1227775554


### Answer

24747430309

## aoc25-2.2
The clerk quickly discovers that there are still invalid IDs in the ranges in your list. Maybe the young Elf was doing other silly patterns as well?

Now, an ID is invalid if it is made only of some sequence of digits repeated at least twice. So, 12341234 (1234 two times), 123123123 (123 three times), 1212121212 (12 five times), and 1111111 (1 seven times) are all invalid IDs.

From the same example as before:

11-22 still has two invalid IDs, 11 and 22.
95-115 now has two invalid IDs, 99 and 111.
998-1012 now has two invalid IDs, 999 and 1010.
1188511880-1188511890 still has one invalid ID, 1188511885.
222220-222224 still has one invalid ID, 222222.
1698522-1698528 still contains no invalid IDs.
446443-446449 still has one invalid ID, 446446.
38593856-38593862 still has one invalid ID, 38593859.
565653-565659 now has one invalid ID, 565656.
824824821-824824827 now has one invalid ID, 824824824.
2121212118-2121212124 now has one invalid ID, 2121212121.
Adding up all the invalid IDs in this example produces 4174379265.

What do you get if you add up all of the invalid IDs using these new rules?

### Input

In [ ]:
path = '/mnt/c/Users/guspa/Desktop/python_homework/data/aoc25/aoc25_2.1.txt'
mode = 'r'

database = []
item = ''

with open(path, mode) as file:
    for line in file.readlines():
        line = line.strip()
        
        for position in range(len(line)):

            if line[position] == ',':
                database.append(item)
                item = ''

            elif position == len(line)-1:
                item += line[position]
                database.append(item)
                item = ''

            else:
                item += line[position]

print(database)

['19391-47353', '9354357-9434558', '4646427538-4646497433', '273-830', '612658-674925', '6639011-6699773', '4426384-4463095', '527495356-527575097', '22323258-22422396', '412175-431622', '492524-611114', '77-122', '992964846-993029776', '165081-338962', '925961-994113', '7967153617-7967231799', '71518058-71542434', '64164836-64292066', '4495586-4655083', '2-17', '432139-454960', '4645-14066', '6073872-6232058', '9999984021-10000017929', '704216-909374', '48425929-48543963', '52767-94156', '26-76', '1252-3919', '123-228']


### Functions

In [ ]:
def get_range(s):
    
    s1 = ''
    s2 = ''
    dash = ''

    for character in range(len(s)):
        if dash == '':
            #print('con1')
            
            if s[character] == '-':
                #print('con1a')
                dash += '-'

            else:
                #print('con1b')
                s1 += s[character]
        
        else:
            #print('con2')
            s2 += s[character]

        if character == len(s)-1:
            #print('clean')
            dash = ''

    return int(s1), int(s2)+1

get_range('11-22') # a, b+1

(11, 23)

In [ ]:
def invalid_repetition(integer): # e.g. 123123123
    # setup variables
    s = str(integer)
    s_length = len(s) # 9
    code_check = ''

    # for loop going through every digit
    for digit in range(s_length): # e.g 0-8

        # buil a sequence that might repeat within an integer
        sequence = int(s[0:(1+digit)]) #e.g 1; 12; 123
        sequence_length = len(str(sequence)) #e.g 1; 2; 3

        # test if a sequence length could divide the length of an integer 
        if s_length % sequence_length != 0: # 0; continue; 0
            continue
        
        # get number of repetitions
        repetitions = s_length // sequence_length #e.g 9, 3

        # skip an integer which has repetition less than 2 (at least 1 repetition)
        if repetitions < 2: #e.g. 9, 3
            continue
        
        # compares integer with a code which is derived as a result of its division by a found sequence
        if integer % sequence == 0: 
            # get result of integer division by a sequence
            code = integer // sequence #e.g 123123123; 1001001
            pattern = '1' + '0' * (sequence_length-1) #e.g. 1; 100
            code_check = pattern * (repetitions-1) + '1' #e.g. 111111111; 1001001

            # return integer once code and built code match
            if str(code) == code_check:
                return integer

        else:
            continue

    return False

invalid_repetition(123123123)

123123123

### Programme

In [ ]:
invalid_ID = 0

for item in database:

    for value in range(get_range(item)[0], get_range(item)[1]):

        if value == invalid_repetition(value):
            invalid_ID += value
        else:
            continue

### Evaluation of Results

In [ ]:
print('sum of invalid IDs:', invalid_ID)

sum of invalid IDs: 30962646823


### Deconstruction of the Example

In [ ]:
# database of examaple ranges
database = ['11-22', '95-115', '998-1012', '1188511880-1188511890', '222220-222224', '1698522-1698528', '446443-446449', '38593856-38593862', '565653-565659', '824824821-824824827', '2121212118-2121212124']

# accumulator variable for sums of invalid ID numbers
invalid_ID = 0

# for each item in databse, do
for item in database:
    # for each value within a range which starts with first part of the ID and ends with the second part of the ID including
    for value in range(get_range(item)[0], get_range(item)[1]):
        # if value is equal to value returned from a function which checks if the value could be invalid, do
        if value == invalid_repetition(value):
            print(value)
            invalid_ID += value # add the value to the accumualator variable
        # is not equal, do
        else:
            continue # go into another iteration (new number within the interval)
    
# print the sum of invalid IDs
print('sum of invalid IDs:', invalid_ID)

11
22
99
111
999
1010
1188511885
222222
446446
38593859
565656
824824824
2121212121
sum of invalid IDs: 4174379265


### Answer
30962646823

## aoc25-3.1
You descend a short staircase, enter the surprisingly vast lobby, and are quickly cleared by the security checkpoint. When you get to the main elevators, however, you discover that each one has a red light above it: they're all offline.

"Sorry about that," an Elf apologizes as she tinkers with a nearby control panel. "Some kind of electrical surge seems to have fried them. I'll try to get them online soon."

You explain your need to get further underground. "Well, you could at least take the escalator down to the printing department, not that you'd get much further than that without the elevators working. That is, you could if the escalator weren't also offline."

"But, don't worry! It's not fried; it just needs power. Maybe you can get it running while I keep working on the elevators."

There are batteries nearby that can supply emergency power to the escalator for just such an occasion. The batteries are each labeled with their joltage rating, a value from 1 to 9. You make a note of their joltage ratings (your puzzle input). For example:

987654321111111
811111111111119
234234234234278
818181911112111
The batteries are arranged into banks; each line of digits in your input corresponds to a single bank of batteries. Within each bank, you need to turn on exactly two batteries; the joltage that the bank produces is equal to the number formed by the digits on the batteries you've turned on. For example, if you have a bank like 12345 and you turn on batteries 2 and 4, the bank would produce 24 jolts. (You cannot rearrange batteries.)

You'll need to find the largest possible joltage each bank can produce. In the above example:

In 987654321111111, you can make the largest joltage possible, 98, by turning on the first two batteries.
In 811111111111119, you can make the largest joltage possible by turning on the batteries labeled 8 and 9, producing 89 jolts.
In 234234234234278, you can make 78 by turning on the last two batteries (marked 7 and 8).
In 818181911112111, the largest joltage you can produce is 92.
The total output joltage is the sum of the maximum joltage from each bank, so in this example, the total output joltage is 98 + 89 + 78 + 92 = 357.

There are many batteries in front of you. Find the maximum joltage possible from each bank; what is the total output joltage?

### Input

In [ ]:
path = '/mnt/c/Users/guspa/Desktop/python_homework/data/aoc25/aoc25_3.1.txt'
mode = 'r'

banks = []

with open(path, mode) as file:
    for line in file.readlines():
        line = line.strip()
        banks.append(line)

print(banks)

['7455337345554393449454442744452533444624555444444525654744644442462265544584444244243377662874573954', '4444433574444444521434318444437434434744463343344862443745783646353544443448334445344547344244354484', '4454325334739566743455434465374426494464323643934468434765454444555535464455456547468454544534444637', '2253633222422222521322222224224222262272124222312722231222222262212233722221212262222125452412392631', '1222111334214231341321332223311232114113234132421133324233143211344224341223243324234444111131256789', '5552574722261445226642632635456622675165425672633256621353613292945452142158642643242263631127224735', '2332532231343123261241313224232361514572322232242233225222222332337363211613222322578362253152228332', '3444583257424284574732453333613283357533643446346353834484734436344577745464538846857954427384448844', '4455353554554545555555435654454545177665544544452544365893276544438552455556355859553169554555553594', '12122132222113253223122324322221223222333221422232324123222222

### Functions

In [ ]:
# none

### Programme

In [ ]:
joltage = 0

for battery_series in banks:

    max1 = None
    max1_past = 0
    max1_position = None

    for position in range(len(battery_series)-1):

        if max1 is None:
            max1 = battery_series[position]
            max1_position = position
            max1_past = max1

        else:
            max1 = max(max1, battery_series[position])
            
            if max1 > max1_past:
                max1_past = max1
                max1_position = position

    max2 = None
    max2_past = 0
    max2_position = None

    for position in range(max1_position+1,len(battery_series)):

        if max2 is None:
            max2 = battery_series[position]
            max2_position = position
            max2_past = max2

        else:
            max2 = max(max2, battery_series[position])
            
            if max2 > max2_past:
                max2_past = max2
                max2_position = position

    series_max = '' + str(max1) + str(max2)
    joltage += int(series_max)

### Evaluation of Results

In [ ]:
print('final joltage is:', joltage)

final joltage is: 17694


### Deconstruction of the Example

In [ ]:
# list with strings as examples
banks = ['987654321111111', '811111111111119', '234234234234278', '818181911112111']
# accumulator variable for sum of "joltages"
joltage = 0

# loop going through all strings in the list
for battery_series in banks:

    # setup variables for max1
    max1 = None
    max1_past = 0
    max1_position = None

    # find max1 from starting digit until the one before the last one
    for position in range(len(battery_series)-1):

        # initial assignemnt of values related to max1
        if max1 is None:
            max1 = battery_series[position]
            max1_position = position
            max1_past = max1

        # initial assignemnt of values related to max1
        else:
            max1 = max(max1, battery_series[position])
            
            #once change is triggered, adjust past value accordingly
            if max1 > max1_past:
                max1_past = max1
                max1_position = position

     # setup variables for max2
    max2 = None
    max2_past = 0
    max2_position = None

    # find max2 from position of max1+1 until the end
    for position in range(max1_position+1,len(battery_series)):

        # initial assignemnt of values related to max2
        if max2 is None:
            max2 = battery_series[position]
            max2_position = position
            max2_past = max2

        # initial assignemnt of values related to max2
        else:
            max2 = max(max2, battery_series[position])
            
            #once change is triggered, adjust past value accordingly
            if max2 > max2_past:
                max2_past = max2
                max2_position = position

    # joltage per series
    series_max = '' + str(max1) + str(max2)
    joltage += int(series_max)

# final joltage
print('final joltage is:', joltage)

final joltage is: 357


### Answer
17694

## aoc25-3.2
The escalator doesn't move. The Elf explains that it probably needs more joltage to overcome the static friction of the system and hits the big red "joltage limit safety override" button. You lose count of the number of times she needs to confirm "yes, I'm sure" and decorate the lobby a bit while you wait.

Now, you need to make the largest joltage by turning on exactly twelve batteries within each bank.

The joltage output for the bank is still the number formed by the digits of the batteries you've turned on; the only difference is that now there will be 12 digits in each bank's joltage output instead of two.

Consider again the example from before:

987654321111111
811111111111119
234234234234278
818181911112111
Now, the joltages are much larger:

In 987654321111111, the largest joltage can be found by turning on everything except some 1s at the end to produce 987654321111.
In the digit sequence 811111111111119, the largest joltage can be found by turning on everything except some 1s, producing 811111111119.
In 234234234234278, the largest joltage can be found by turning on everything except a 2 battery, a 3 battery, and another 2 battery near the start to produce 434234234278.
In 818181911112111, the joltage 888911112111 is produced by turning on everything except some 1s near the front.
The total output joltage is now much larger: 987654321111 + 811111111119 + 434234234278 + 888911112111 = 3121910778619.

What is the new total output joltage?

### Input

In [ ]:
path = '/mnt/c/Users/guspa/Desktop/python_homework/data/aoc25/aoc25_3.1.txt'
mode = 'r'

banks = []

with open(path, mode) as file:
    for line in file.readlines():
        line = line.strip()
        banks.append(line)

print(banks)

['7455337345554393449454442744452533444624555444444525654744644442462265544584444244243377662874573954', '4444433574444444521434318444437434434744463343344862443745783646353544443448334445344547344244354484', '4454325334739566743455434465374426494464323643934468434765454444555535464455456547468454544534444637', '2253633222422222521322222224224222262272124222312722231222222262212233722221212262222125452412392631', '1222111334214231341321332223311232114113234132421133324233143211344224341223243324234444111131256789', '5552574722261445226642632635456622675165425672633256621353613292945452142158642643242263631127224735', '2332532231343123261241313224232361514572322232242233225222222332337363211613222322578362253152228332', '3444583257424284574732453333613283357533643446346353834484734436344577745464538846857954427384448844', '4455353554554545555555435654454545177665544544452544365893276544438552455556355859553169554555553594', '12122132222113253223122324322221223222333221422232324123222222

### Functions

In [ ]:
def max_pos(l):

    maximum = max(l)
    maximum_position = l.index(maximum)

    return maximum, int(maximum_position)

max_pos('987654321111111')

('9', 0)

### Programme
#### Failed Attempt

In [ ]:
banks = ['987654321111111', '811111111111119', '234234234234278', '818181911112111']

joltage = 0

for battery_series in banks:

    l_battery_series = list(battery_series)
    pattern = list('0' * len(battery_series))

    for battery in range(12):

        maximum, maximum_position = max_pos(l_battery_series)

        pattern[maximum_position] = maximum
        l_battery_series[maximum_position] = '0'

    while '0' in pattern:
        pattern.remove('0')
    print(pattern)

    joltage = int(''.join(pattern))
    print(joltage)

['9', '8', '7', '6', '5', '4', '3', '2', '1', '1', '1', '1']
987654321111
['8', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '9']
811111111119
['2', '3', '4', '2', '3', '4', '3', '4', '3', '4', '7', '8']
234234343478
['8', '1', '8', '1', '8', '1', '9', '1', '1', '1', '1', '2']
818181911112


#### Found Solution
Compares new with previous, which was used to build the pattern, whether the new is larger, if yes, the previous in the pattern is removed. This is repeated until 3 deletions occur, resulting in the remaining number we need.

In [ ]:
joltage = 0
banks = ['987654321111111', '811111111111119', '234234234234278', '818181911112111']

for battery_series in banks:
    pattern = []
    remove_max = len(battery_series) - 12
    for ch_digit in battery_series:
        while remove_max > 0 and pattern != [] and pattern[-1] < ch_digit:
            pattern.pop()
            remove_max -= 1
        pattern.append(ch_digit)
    joltage += int(''.join(pattern[:12]))
print(joltage)

#### Successful Attempt

In [ ]:
joltage = 0

for battery_series in banks:

    l_battery_series = list(battery_series)

    batteries = 12
    start = 0
    pattern = []

    while batteries > 0:
        end = len(l_battery_series) - (batteries - 1)

        check = l_battery_series[start:end] 

        maximum, maximum_position = max_pos(check)
        maximum_position += start
        start = maximum_position + 1

        pattern.append(maximum)

        batteries -= 1

    joltage += int(''.join(pattern))

175659236361660


### Evaluation of Results

In [ ]:
print('total necessary joltage:', joltage)

total necessary joltage: 3121910778619


### Deconstruction of the Examples

In [ ]:
# list with strings as examples
banks = ['987654321111111', '811111111111119', '234234234234278', '818181911112111']
# accumulator variable for sum of "joltages"
joltage = 0

# loop going through all strings in the list
for battery_series in banks:

    l_battery_series = list(battery_series) # transfromation of a string into a list

    batteries = 12 # number of necessary batteries
    start = 0 # starting position
    pattern = [] # accumulator variable for building a pattern

    # while loop going until a 12 digit number is found
    while batteries > 0:
  
        end = len(l_battery_series) - (batteries - 1) # end position until which search takes place; -12 at the beginning to allow for 12 positions to be filled; -1 to allow for movement towards right

        check = l_battery_series[start:end] # search window with size adjusted according to start and stop, but only start position narrows the window; both start and end are adjusted to progress towards right by +1 (start, from starting position) -1 (end, from the last position)

        maximum, maximum_position = max_pos(check) # function returning maximum and its position/index
        maximum_position += start # add starting position
        start = maximum_position + 1 # prepare start position for next iteration by becoming maximum_position (with start position) + 1 (to move by 1 towards right)

        pattern.append(maximum) # maximum is appended into the pattern variable

        batteries -= 1 # makes sure that the window moves by one towards the end

    # construct joltage integer from the pattern
    joltage += int(''.join(pattern))

print('total necessary joltage:', joltage)

# check first 4 at the beginning, starting from the left, because if 3 before were wrong at the beginning then the 4th until the 15th from the 15 digit number would be the number we need; the check is adjusted, if numbers are skipped at the beginning, the search is narrowed accordingly - otherwise 4 positions remain being checked with position moved by 1 to the right

joltage: 3121910778619


### Answer
175659236361660

## aoc25-4.1
You ride the escalator down to the printing department. They're clearly getting ready for Christmas; they have lots of large rolls of paper everywhere, and there's even a massive printer in the corner (to handle the really big print jobs).

Decorating here will be easy: they can make their own decorations. What you really need is a way to get further into the North Pole base while the elevators are offline.

"Actually, maybe we can help with that," one of the Elves replies when you ask for help. "We're pretty sure there's a cafeteria on the other side of the back wall. If we could break through the wall, you'd be able to keep moving. It's too bad all of our forklifts are so busy moving those big rolls of paper around."

If you can optimize the work the forklifts are doing, maybe they would have time to spare to break through the wall.

The rolls of paper (@) are arranged on a large grid; the Elves even have a helpful diagram (your puzzle input) indicating where everything is located.

For example:

..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.

The forklifts can only access a roll of paper if there are fewer than four rolls of paper in the eight adjacent positions. If you can figure out which rolls of paper the forklifts can access, they'll spend less time looking and more time breaking down the wall to the cafeteria.

In this example, there are 13 rolls of paper that can be accessed by a forklift (marked with x):

..xx.xx@x.
x@@.@.@.@@
@@@@@.x.@@
@.@@@@..@.
x@.@@@@.@x
.@@@@@@@.@
.@.@.@.@@@
x.@@@.@@@@
.@@@@@@@@.
x.x.@@@.x.
Consider your complete diagram of the paper roll locations. How many rolls of paper can be accessed by a forklift?

## Packages

In [ ]:
# for dataframe structure
import pandas as pd

# test
data = {'Name': ['Alice', 'Bob', 'Charlie'],'Age': [25, 30, 35]}
df = pd.DataFrame(data)
print(df)

print()

df_transposed = df.transpose()
print(df_transposed)

      Name  Age
0    Alice   25
1      Bob   30
2  Charlie   35

          0    1        2
Name  Alice  Bob  Charlie
Age      25   30       35


## Input

In [ ]:
path = '/mnt/c/Users/guspa/Desktop/python_homework/data/aoc25/aoc25_4.1.txt'
mode = 'r'

diagram_dic = {} # dictionary setup for a dataframe

# count lines first
with open(path, mode) as file:
    line_count = sum(1 for line in file)

line_count_bounded = line_count + 2 # we need to add 2 to account for 2 added columns with "dot" characters

with open(path, mode) as file:
    # insert empty column before 1st line
    diagram_dic.update({0: list('.' * (line_count_bounded))})

    for column, line in enumerate(file):
        line = line.strip()
        line = '.' + line + '.'
        diagram_dic.update({column + 1: list(line)}) # start from 2nd column
    
    # insert empty column after last line
    diagram_dic.update({column + 2: list('.' * (line_count_bounded))})

diagram_df = pd.DataFrame(diagram_dic) # dataframe
diagram_cr = diagram_df.transpose() # transposed dataframe - working version, since otherwise a signle dictionary input is considered as a single column and we need the input in the position of rows
print(diagram_cr)

    0   1   2   3   4   5   6   7   8   9    ... 129 130 131 132 133 134 135  \
0     .   .   .   .   .   .   .   .   .   .  ...   .   .   .   .   .   .   .   
1     .   @   @   @   @   @   .   @   @   @  ...   @   .   .   @   .   @   @   
2     .   .   .   .   @   @   @   .   @   @  ...   @   .   @   .   @   .   .   
3     .   @   @   .   @   @   @   @   @   @  ...   .   @   .   .   .   @   @   
4     .   .   @   @   @   @   .   @   @   .  ...   .   @   .   @   @   @   @   
..   ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ..  ..  ..  ..  ..  ..  ..   
134   .   .   @   @   @   @   @   .   @   .  ...   .   @   .   @   .   @   @   
135   .   .   @   @   .   @   @   @   .   @  ...   @   .   @   .   .   .   .   
136   .   @   @   .   @   @   @   .   @   @  ...   .   @   .   @   .   @   @   
137   .   @   .   .   @   @   @   @   .   @  ...   @   @   @   @   .   .   @   
138   .   .   .   .   .   .   .   .   .   .  ...   .   .   .   .   .   .   .   

    136 137 138  
0     .   .   .  
1  

## Testing Ground

In [ ]:
# ARTIFICIAL
# test_dic = {0:['@','@','.','@','@'],
#             1:['.','@','@','@','@'],
#             2:['@','@','@','@','@'],
#             3:['@','.','.','@','.'],
#             4:['.','@','@','@','@']}

# test_dic = {0:['.','.','.','.','.','.','.'],
#             1:['.','@','@','.','@','@','.'],
#             2:['.','.','@','@','@','@','.'],
#             3:['.','@','@','@','@','@','.'],
#             4:['.','@','.','.','@','.','.'],
#             5:['.','.','@','@','@','@','.'],
#             6:['.','.','.','.','.','.','.']}

# test_df = pd.DataFrame(test_dic)

# test_cr = test_df.transpose()
# print(test_cr)

#print(test_cr[0][1]) # columns and rows

# PROVIDED

test_path = '/mnt/c/Users/guspa/Desktop/python_homework/data/aoc25/test_example_aoc25_4.1.txt'
test_mode = 'r'

test_diagram_dic = {} # dictionary setup for a dataframe

# count lines first
with open(test_path, test_mode) as test_file:
    test_line_count = sum(1 for line in test_file)

test_line_count_bounded = test_line_count + 2 # we need to add 2 to account for 2 added columns with "dot" characters

with open(test_path, test_mode) as test_file:
    # insert empty column before 1st line
    test_diagram_dic.update({0: list('.' * test_line_count_bounded)})

    for test_column, test_line in enumerate(test_file):
        test_line = test_line.strip()
        test_line = '.' + test_line + '.'
        test_diagram_dic.update({test_column + 1: list(test_line)}) # start from 2nd column
    
    # insert empty column after last line
    test_diagram_dic.update({test_column + 2: list('.' * (test_line_count_bounded))})

test_diagram_df = pd.DataFrame(test_diagram_dic) # dataframe
test_diagram_cr = test_diagram_df.transpose() # transposed dataframe - working version, since otherwise a signle dictionary input is considered as a single column and we need the input in the position of rows
print(test_diagram_cr)

#print()
#print(test_diagram_cr[0][1]) # columns and rows

   0  1  2  3  4  5  6  7  8  9  10 11
0   .  .  .  .  .  .  .  .  .  .  .  .
1   .  .  .  @  @  .  @  @  @  @  .  .
2   .  @  @  @  .  @  .  @  .  @  @  .
3   .  @  @  @  @  @  .  @  .  @  @  .
4   .  @  .  @  @  @  @  .  .  @  .  .
5   .  @  @  .  @  @  @  @  .  @  @  .
6   .  .  @  @  @  @  @  @  @  .  @  .
7   .  .  @  .  @  .  @  .  @  @  @  .
8   .  @  .  @  @  @  .  @  @  @  @  .
9   .  .  @  @  @  @  @  @  @  @  .  .
10  .  @  .  @  .  @  @  @  .  @  .  .
11  .  .  .  .  .  .  .  .  .  .  .  .


## Functions

In [ ]:
def scan(position,  dataframe):

    pos_c = position[0]
    pos_r = position[1]

    surrounding_rolls = 0

    if dataframe[pos_c][pos_r] == '@':

        for c in range(-1, 2): # -1, 0, 1
            for r in range(-1, 2): # -1, 0, 1

                if r == 0 and c == 0: # skip centre
                    continue
                
                check_c = pos_c + c
                check_r = pos_r + r
                
                if 0 <= check_r < len(dataframe) and 0 <= check_c < len(dataframe):
                    if dataframe[check_c][check_r] == '@':
                        surrounding_rolls += 1
        
        if surrounding_rolls < 4:
            return True
        else:
            return False
    
    return False

scan([3, 1], test_diagram_cr) # c=3, r=1

True

## Programme

In [ ]:
accessible_rolls = 0

for column in range(len(diagram_cr.columns)):
     for row in range(len(diagram_cr)):
            
            if scan([column, row], diagram_cr):
                   accessible_rolls += 1

            else: 
                continue

## Evaluation of Results

In [ ]:
print('number of accessible rolls:', accessible_rolls)

number of accessible rolls: 1502


## Deconstruction of the Example

In [ ]:
# integer based accumulator variable for rolls that are accessible - counter
accessible_rolls = 0

# go through each column of a df
for column in range(len(test_diagram_cr.columns)):
       # go through each row of a df
       for row in range(len(test_diagram_cr)):

              pos_c = column # save current column position
              pos_r = row # save current row position
              surrounding_rolls = 0 # accumulator variable for count of surrounding rolls

              # check if there is "@" at the current position
              if test_diagram_cr[pos_c][pos_r] == '@':

                     # check surroundings in 3x3 with current position in the middle
                     for c in range(-1, 2): # -1, 0, 1
                            for r in range(-1, 2): # -1, 0, 1
                                   
                                   if r == 0 and c == 0: # skip centre, the current position
                                          continue
                                   
                                   # add positions to current positions from the loops
                                   check_c = pos_c + c
                                   check_r = pos_r + r

                                   # if the position is within the dataframe itself (not outside)
                                   if 0 <= check_r < len(test_diagram_cr) and 0 <= check_c < len(test_diagram_cr):
                                          
                                          # if there is "@" at one of the positions, add 1 to counter of surrounding rolls
                                          if test_diagram_cr[check_c][check_r] == '@':
                                                 surrounding_rolls += 1
                     
                     # consider the roll to be accessible if there is less than 4 rolls in the surroundings of 8 position
                     if surrounding_rolls < 4:
                                   # add 1 to counter of accessible rolls
                                   accessible_rolls += 1

                     else:
                            continue

# print the number of all easily accessible rolls
print('number of accessible rolls:', accessible_rolls)

number of accessible rolls: 13


## aoc-4.2
Now, the Elves just need help accessing as much of the paper as they can.

Once a roll of paper can be accessed by a forklift, it can be removed. Once a roll of paper is removed, the forklifts might be able to access more rolls of paper, which they might also be able to remove. How many total rolls of paper could the Elves remove if they keep repeating this process?

Starting with the same example as above, here is one way you could remove as many rolls of paper as possible, using highlighted @ to indicate that a roll of paper is about to be removed, and using x to indicate that a roll of paper was just removed:

Initial state:
..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.

Remove 13 rolls of paper:
..xx.xx@x.
x@@.@.@.@@
@@@@@.x.@@
@.@@@@..@.
x@.@@@@.@x
.@@@@@@@.@
.@.@.@.@@@
x.@@@.@@@@
.@@@@@@@@.
x.x.@@@.x.

Remove 12 rolls of paper:
.......x..
.@@.x.x.@x
x@@@@...@@
x.@@@@..x.
.@.@@@@.x.
.x@@@@@@.x
.x.@.@.@@@
..@@@.@@@@
.x@@@@@@@.
....@@@...

Remove 7 rolls of paper:
..........
.x@.....x.
.@@@@...xx
..@@@@....
.x.@@@@...
..@@@@@@..
...@.@.@@x
..@@@.@@@@
..x@@@@@@.
....@@@...

Remove 5 rolls of paper:
..........
..x.......
.x@@@.....
..@@@@....
...@@@@...
..x@@@@@..
...@.@.@@.
..x@@.@@@x
...@@@@@@.
....@@@...

Remove 2 rolls of paper:
..........
..........
..x@@.....
..@@@@....
...@@@@...
...@@@@@..
...@.@.@@.
...@@.@@@.
...@@@@@x.
....@@@...

Remove 1 roll of paper:
..........
..........
...@@.....
..x@@@....
...@@@@...
...@@@@@..
...@.@.@@.
...@@.@@@.
...@@@@@..
....@@@...

Remove 1 roll of paper:
..........
..........
...x@.....
...@@@....
...@@@@...
...@@@@@..
...@.@.@@.
...@@.@@@.
...@@@@@..
....@@@...

Remove 1 roll of paper:
..........
..........
....x.....
...@@@....
...@@@@...
...@@@@@..
...@.@.@@.
...@@.@@@.
...@@@@@..
....@@@...

Remove 1 roll of paper:
..........
..........
..........
...x@@....
...@@@@...
...@@@@@..
...@.@.@@.
...@@.@@@.
...@@@@@..
....@@@...
Stop once no more rolls of paper are accessible by a forklift. In this example, a total of 43 rolls of paper can be removed.

Start with your original diagram. How many rolls of paper in total can be removed by the Elves and their forklifts?

## Packages

In [ ]:
# for dataframe structure
import pandas as pd

# test
data = {'Name': ['Alice', 'Bob', 'Charlie'],'Age': [25, 30, 35]}
df = pd.DataFrame(data)
print(df)

print()

df_transposed = df.transpose()
print(df_transposed)

      Name  Age
0    Alice   25
1      Bob   30
2  Charlie   35

          0    1        2
Name  Alice  Bob  Charlie
Age      25   30       35


## Input

In [ ]:
path = '/mnt/c/Users/guspa/Desktop/python_homework/data/aoc25/aoc25_4.1.txt'
mode = 'r'

diagram_dic = {} # dictionary setup for a dataframe

# count lines first
with open(path, mode) as file:
    line_count = sum(1 for line in file)

line_count_bounded = line_count + 2 # we need to add 2 to account for 2 added columns with "dot" characters

with open(path, mode) as file:
    # insert empty column before 1st line
    diagram_dic.update({0: list('.' * (line_count_bounded))})

    for column, line in enumerate(file):
        line = line.strip()
        line = '.' + line + '.'
        diagram_dic.update({column + 1: list(line)}) # start from 2nd column
    
    # insert empty column after last line
    diagram_dic.update({column + 2: list('.' * (line_count_bounded))})

diagram_df = pd.DataFrame(diagram_dic) # dataframe
diagram_cr = diagram_df.transpose() # transposed dataframe - working version, since otherwise a signle dictionary input is considered as a single column and we need the input in the position of rows
print(diagram_cr)

    0   1   2   3   4   5   6   7   8   9    ... 129 130 131 132 133 134 135  \
0     .   .   .   .   .   .   .   .   .   .  ...   .   .   .   .   .   .   .   
1     .   @   @   @   @   @   .   @   @   @  ...   @   .   .   @   .   @   @   
2     .   .   .   .   @   @   @   .   @   @  ...   @   .   @   .   @   .   .   
3     .   @   @   .   @   @   @   @   @   @  ...   .   @   .   .   .   @   @   
4     .   .   @   @   @   @   .   @   @   .  ...   .   @   .   @   @   @   @   
..   ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ..  ..  ..  ..  ..  ..  ..   
134   .   .   @   @   @   @   @   .   @   .  ...   .   @   .   @   .   @   @   
135   .   .   @   @   .   @   @   @   .   @  ...   @   .   @   .   .   .   .   
136   .   @   @   .   @   @   @   .   @   @  ...   .   @   .   @   .   @   @   
137   .   @   .   .   @   @   @   @   .   @  ...   @   @   @   @   .   .   @   
138   .   .   .   .   .   .   .   .   .   .  ...   .   .   .   .   .   .   .   

    136 137 138  
0     .   .   .  
1  

## Testing Ground

In [ ]:
# ARTIFICIAL
# test_dic = {0:['@','@','.','@','@'],
#             1:['.','@','@','@','@'],
#             2:['@','@','@','@','@'],
#             3:['@','.','.','@','.'],
#             4:['.','@','@','@','@']}

# test_dic = {0:['.','.','.','.','.','.','.'],
#             1:['.','@','@','.','@','@','.'],
#             2:['.','.','@','@','@','@','.'],
#             3:['.','@','@','@','@','@','.'],
#             4:['.','@','.','.','@','.','.'],
#             5:['.','.','@','@','@','@','.'],
#             6:['.','.','.','.','.','.','.']}

# test_df = pd.DataFrame(test_dic)

# test_cr = test_df.transpose()
# print(test_cr)

#print(test_cr[0][1]) # columns and rows

# PROVIDED
test_path = '/mnt/c/Users/guspa/Desktop/python_homework/data/aoc25/test_example_aoc25_4.1.txt'
test_mode = 'r'

test_diagram_dic = {} # dictionary setup for a dataframe

# count lines first
with open(test_path, test_mode) as test_file:
    test_line_count = sum(1 for line in test_file)

test_line_count_bounded = test_line_count + 2 # we need to add 2 to account for 2 added columns with "dot" characters

with open(test_path, test_mode) as test_file:
    # insert empty column before 1st line
    test_diagram_dic.update({0: list('.' * test_line_count_bounded)})

    for test_column, test_line in enumerate(test_file):
        test_line = test_line.strip()
        test_line = '.' + test_line + '.'
        test_diagram_dic.update({test_column + 1: list(test_line)}) # start from 2nd column
    
    # insert empty column after last line
    test_diagram_dic.update({test_column + 2: list('.' * (test_line_count_bounded))})

test_diagram_df = pd.DataFrame(test_diagram_dic) # dataframe
test_diagram_cr = test_diagram_df.transpose() # transposed dataframe - working version, since otherwise a signle dictionary input is considered as a single column and we need the input in the position of rows
print(test_diagram_cr)

#print()
#print(test_diagram_cr[0][1]) # columns and rows

   0  1  2  3  4  5  6  7  8  9  10 11
0   .  .  .  .  .  .  .  .  .  .  .  .
1   .  .  .  @  @  .  @  @  @  @  .  .
2   .  @  @  @  .  @  .  @  .  @  @  .
3   .  @  @  @  @  @  .  @  .  @  @  .
4   .  @  .  @  @  @  @  .  .  @  .  .
5   .  @  @  .  @  @  @  @  .  @  @  .
6   .  .  @  @  @  @  @  @  @  .  @  .
7   .  .  @  .  @  .  @  .  @  @  @  .
8   .  @  .  @  @  @  .  @  @  @  @  .
9   .  .  @  @  @  @  @  @  @  @  .  .
10  .  @  .  @  .  @  @  @  .  @  .  .
11  .  .  .  .  .  .  .  .  .  .  .  .


## Functions

In [ ]:
def scan(position,  dataframe):

    pos_c = position[0]
    pos_r = position[1]

    surrounding_rolls = 0

    if dataframe[pos_c][pos_r] == '@':

        for c in range(-1, 2): # -1, 0, 1
            for r in range(-1, 2): # -1, 0, 1

                if r == 0 and c == 0: # skip centre
                    continue
                
                check_c = pos_c + c
                check_r = pos_r + r
                
                if 0 <= check_r < len(dataframe) and 0 <= check_c < len(dataframe):
                    if dataframe[check_c][check_r] == '@':
                        surrounding_rolls += 1
        
        if surrounding_rolls < 4:
            return True
        else:
            return False
    
    return False

scan([3, 1], test_diagram_cr) # c=3, r=1

True

## Programme

In [ ]:
live_diagram_cr = diagram_cr.copy() # test_diagram_cr or diagmra_cr
accessible_rolls = 0

while True:

    positions_to_change = []

    for column in range(len(live_diagram_cr.columns)):
        for row in range(len(live_diagram_cr)):
            
            if scan([column, row], live_diagram_cr) == True:
                    accessible_rolls += 1
                    positions_to_change.append([column, row])
            else: 
                continue

    if not positions_to_change: # if the list is empty, stop
        break

    for position in range(len(positions_to_change)):
        live_diagram_cr.at[positions_to_change[position][1], positions_to_change[position][0]] = 'x'

## Evaluation of Results

In [ ]:
print('number of accessible rolls:', accessible_rolls)

number of accessible rolls: 13


## Deconstruction of the Example

In [ ]:
# get a copy of original, but adjusted with borders, diagram to work with
live_diagram_cr = diagram_cr.copy() # test_diagram_cr or diagmra_cr
# integer based accumulator variable for rolls that are accessible - counter
accessible_rolls = 0

# start the while loop
while True:

       # empty list accumulator variable to prevent accumualtion of coordinates being saved
       positions_to_change = []

       # go through each column of a df
       for column in range(len(live_diagram_cr.columns)):
              # go through each row of a df
              for row in range(len(live_diagram_cr)):

                     pos_c = column # save current column position
                     pos_r = row # save current row position
                     surrounding_rolls = 0 # accumulator variable for count of surrounding rolls

                     # check if there is "@" at the current position
                     if live_diagram_cr[pos_c][pos_r] == '@':

                            # check surroundings in 3x3 with current position in the middle
                            for c in range(-1, 2): # -1, 0, 1
                                   for r in range(-1, 2): # -1, 0, 1
                                   
                                          if r == 0 and c == 0: # skip centre, the current position
                                                 continue
                                   
                                          # add positions to current positions from the loops
                                          check_c = pos_c + c
                                          check_r = pos_r + r

                                          # if the position is within the dataframe itself (not outside)
                                          if 0 <= check_r < len(live_diagram_cr) and 0 <= check_c < len(live_diagram_cr):
                                          
                                                 # if there is "@" at one of the positions, add 1 to counter of surrounding rolls
                                                 if live_diagram_cr[check_c][check_r] == '@':
                                                        surrounding_rolls += 1
                     
                            # consider the roll to be accessible if there is less than 4 rolls in the surroundings of 8 position
                            if surrounding_rolls < 4:
                                   # add 1 to counter of easily accessible rolls
                                   accessible_rolls += 1
                                   # add coordinates whenever an accessible roll, which has less than 4 rolls in its surroudnings, is found 
                                   positions_to_change.append([column, row])

       # if the list with coordinates is empty (no more new coordinates), stop
       if not positions_to_change:
              break
       
       # for each coordinates, change the symbol of easily accessible roll to "x"
       for position in range(len(positions_to_change)):
              live_diagram_cr.at[positions_to_change[position][1], positions_to_change[position][0]] = 'x'

# print the number of all easily accessible rolls
print('number of accessible rolls:', accessible_rolls)

number of accessible rolls: 9083


## Answer
9083

## aoc25-5.1
As the forklifts break through the wall, the Elves are delighted to discover that there was a cafeteria on the other side after all.

You can hear a commotion coming from the kitchen. "At this rate, we won't have any time left to put the wreaths up in the dining hall!" Resolute in your quest, you investigate.

"If only we hadn't switched to the new inventory management system right before Christmas!" another Elf exclaims. You ask what's going on.

The Elves in the kitchen explain the situation: because of their complicated new inventory management system, they can't figure out which of their ingredients are fresh and which are spoiled. When you ask how it works, they give you a copy of their database (your puzzle input).

The database operates on ingredient IDs. It consists of a list of fresh ingredient ID ranges, a blank line, and a list of available ingredient IDs. For example:

3-5
10-14
16-20
12-18

1
5
8
11
17
32
The fresh ID ranges are inclusive: the range 3-5 means that ingredient IDs 3, 4, and 5 are all fresh. The ranges can also overlap; an ingredient ID is fresh if it is in any range.

The Elves are trying to determine which of the available ingredient IDs are fresh. In this example, this is done as follows:

Ingredient ID 1 is spoiled because it does not fall into any range.
Ingredient ID 5 is fresh because it falls into range 3-5.
Ingredient ID 8 is spoiled.
Ingredient ID 11 is fresh because it falls into range 10-14.
Ingredient ID 17 is fresh because it falls into range 16-20 as well as range 12-18.
Ingredient ID 32 is spoiled.
So, in this example, 3 of the available ingredient IDs are fresh.

Process the database file from the new inventory management system. How many of the available ingredient IDs are fresh?

## Input

In [11]:
path = "/mnt/c/users/guspa/Desktop/python_homework/data/aoc25/aoc25_5.1.txt"
mode = "r"

database = []

with open(path, mode) as file:
    for line in file.readlines():
        line = line.strip()
        database.append(line)

print(database)

['305611713170986-307905965847624', '12998381660604-15936296075106', '162066655737986-166911023643586', '510900132568076-512313213628552', '706001751286-1356033568084', '426331119734793-426918262201595', '232729491567558-240799959747013', '425520625269696-426203588375521', '268589839302465-268697910083944', '352543118293729-353012217438390', '430127896135595-430376089286127', '265050584829564-265842337112102', '93469731113983-96965898234540', '514212111806029-514212111806029', '275174069600783-280407674989496', '429286175409489-429590051038394', '524376152964524-531669803105338', '294164413666790-294164413666790', '61167866556730-68484069473264', '504055856478078-505668530691037', '382509364383477-389403596739223', '162066655737986-166911023643586', '270774761986680-271062124264236', '195203568087763-198855569516936', '430600880857969-430872135293366', '353012217438390-353194043731516', '422616158235083-423267877109087', '338366511702246-339908568671883', '5702068752562-6212012481365',

## Testing Ground

In [ ]:
database = ["3-5", 
            "10-15", 
            "16-20", 
            "12-18", 
            "" , 
            "1", 
            "5", 
            "8", 
            "11", 
            "17", 
            "32"]

print(database)

['3-5', '10-15', '16-20', '12-18', ' ', '1', '5', '8', '11', '17', '32']


## Functions

In [ ]:
def separate_lists(l):
    ranges = []
    integers = []

    for string in l:

        if "-" in string:
            num1 = int(string.split("-")[0])
            num2 = int(string.split("-")[1])
            ranges.append([num1, num2])
        
        elif string == "":
            continue

        else:
            integers.append(int(string))
    
    return ranges, integers

separate_lists(database)

([[305611713170986, 307905965847624],
  [12998381660604, 15936296075106],
  [162066655737986, 166911023643586],
  [510900132568076, 512313213628552],
  [706001751286, 1356033568084],
  [426331119734793, 426918262201595],
  [232729491567558, 240799959747013],
  [425520625269696, 426203588375521],
  [268589839302465, 268697910083944],
  [352543118293729, 353012217438390],
  [430127896135595, 430376089286127],
  [265050584829564, 265842337112102],
  [93469731113983, 96965898234540],
  [514212111806029, 514212111806029],
  [275174069600783, 280407674989496],
  [429286175409489, 429590051038394],
  [524376152964524, 531669803105338],
  [294164413666790, 294164413666790],
  [61167866556730, 68484069473264],
  [504055856478078, 505668530691037],
  [382509364383477, 389403596739223],
  [162066655737986, 166911023643586],
  [270774761986680, 271062124264236],
  [195203568087763, 198855569516936],
  [430600880857969, 430872135293366],
  [353012217438390, 353194043731516],
  [422616158235083, 423

In [3]:
def comparison(start, stop, integer):
    if integer >= start and integer <= stop:
        return True
    else:
        return False

comparison(3, 5, 4)

True

## Programme

In [12]:
ingredient_ranges, available_ingredients = separate_lists(database)

fresh = 0

for ingredient in available_ingredients:

    for range in ingredient_ranges:

        start = range[0]
        stop = range[1]

        if comparison(start, stop, ingredient) == True:
            fresh += 1
            break

## Evaluation of Results

In [13]:
print('count of fresh ingredients:', fresh)

count of fresh ingredients: 773


## Deconstruction of the Example

In [14]:
# example list containing the ingredient ID ranges, stripped empty space and available ingredient IDs
database = ["3-5", "10-15", "16-20", "12-18", "" , "1", "5", "8", "11", "17", "32"]
# list accumulator variable for extracted ID ranges
ingredient_ranges = []
# list accumulator variable for extracted available ingredient IDs
available_ingredients = []
# integer accumulator variable for fresh ingredients
fresh = 0

# for each string database
for string in database:
    
    # does the string contain "-"?
    if "-" in string:
        num1 = int(string.split("-")[0]) # num1 becomes an integer, the first part before "-"
        num2 = int(string.split("-")[1]) # num2 becomes an integer, the second part after "-"
        ingredient_ranges.append([num1, num2]) # add num1 and num2 as a list into "ranges" variable

    # does the string contain ""?    
    elif string == "":
        continue # skip

    # otehrwise
    else: 
        available_ingredients.append(int(string)) # add numbers as integers into "integers" variable


# for each ingredient in a list of "available_ingredients"
for ingredient in available_ingredients:

    # for each range in a list of "ingredient_ranges"
    for range in ingredient_ranges:
        start = range[0] # get first number from the range
        stop = range[1] # get last numebr from the range

        # if ingredient ID is within the interval (first and last numbers)
        if ingredient >= start and ingredient <= stop:
            fresh += 1 # add 1 to counter of fresh ingredients
            break # break out of the loop and continue with iteration through the list of "available_ingredients" once found for the first time

# print the count of fresh ingredients
print('count of fresh ingredients:', fresh)

count of fresh ingredients: 3


## Answer
773

## aoc25-5.2
The Elves start bringing their spoiled inventory to the trash chute at the back of the kitchen.

So that they can stop bugging you when they get new inventory, the Elves would like to know all of the IDs that the fresh ingredient ID ranges consider to be fresh. An ingredient ID is still considered fresh if it is in any range.

Now, the second section of the database (the available ingredient IDs) is irrelevant. Here are the fresh ingredient ID ranges from the above example:

3-5
10-14
16-20
12-18
The ingredient IDs that these ranges consider to be fresh are 3, 4, 5, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, and 20. So, in this example, the fresh ingredient ID ranges consider a total of 14 ingredient IDs to be fresh.

Process the database file again. How many ingredient IDs are considered to be fresh according to the fresh ingredient ID ranges?

## Input

In [242]:
path = "/mnt/c/users/guspa/Desktop/python_homework/data/aoc25/aoc25_5.1.txt"
mode = "r"

ranges_provided = []

with open(path, mode) as file:
    for line in file.readlines():
        line = line.strip()

        if line == "":
            break
        else:
            num1 = int(line.split("-")[0]) # take the first number of the interval
            num2 = int(line.split("-")[1]) # take the second number of the interval
            ranges_provided.append([num1, num2]) # store them as a list

print(ranges_provided)

[[305611713170986, 307905965847624], [12998381660604, 15936296075106], [162066655737986, 166911023643586], [510900132568076, 512313213628552], [706001751286, 1356033568084], [426331119734793, 426918262201595], [232729491567558, 240799959747013], [425520625269696, 426203588375521], [268589839302465, 268697910083944], [352543118293729, 353012217438390], [430127896135595, 430376089286127], [265050584829564, 265842337112102], [93469731113983, 96965898234540], [514212111806029, 514212111806029], [275174069600783, 280407674989496], [429286175409489, 429590051038394], [524376152964524, 531669803105338], [294164413666790, 294164413666790], [61167866556730, 68484069473264], [504055856478078, 505668530691037], [382509364383477, 389403596739223], [162066655737986, 166911023643586], [270774761986680, 271062124264236], [195203568087763, 198855569516936], [430600880857969, 430872135293366], [353012217438390, 353194043731516], [422616158235083, 423267877109087], [338366511702246, 339908568671883], [5

## Testing Ground

In [228]:
test_ranges = [[3,5], 
               [10,15], 
               [16,20], 
               [12,18]]

print(test_ranges)

[[3, 5], [10, 15], [16, 20], [12, 18]]


## Functions

In [229]:
def check_merge(range1, range2):

    i = range1[0]
    j = range1[1]

    k = range2[0]
    l = range2[1]
    
    if i <= k and k <= j+1:
        
        merged = [min(i, k), max(j, l)]
        return merged

    elif k <= i and i <= l+1:

        merged = [min(i, k), max(j, l)]       
        return merged
    
    else:
        return False

print(check_merge(test_ranges[1], test_ranges[2]))

[10, 20]


## Programme

In [ ]:
merged = []
#ranges = test_ranges.copy()
ranges = ranges_provided.copy()

while ranges:

    if len(ranges) == 1:
        merged.append(range)
        ranges.remove(range)
    
    for position, range in enumerate(ranges):

        if position == 0:
            current = range

        else:

            check_result = check_merge(current, range)

            if check_result != False:

                ranges.append(check_result)

                ranges.remove(current)
                ranges.remove(range)
                break

            else:
                
                if position == len(ranges)-1:

                    merged.append(current)
                    ranges.remove(current)

                continue

fresh_ingredients = 0      

for first, last in merged:
    fresh_ingredients += last - first + 1     

## Evaluation of Results

In [250]:
print("Total number of ingredients to be considered fresh:", fresh_ingredients)

Total number of ingredients to be considered fresh: 332067203034711


## Deconstruction of the Example

In [ ]:
# list accumulator variable
merged = []
# provided list of example ID ranges
ranges = test_ranges.copy()

# provided list of ID ranges
#ranges = ranges_provided.copy()

# until "ranges" list variable is empty, do
while ranges:

    # when there is the last range in the list of fresh ingredient ID ranges
    if len(ranges) == 1: 
        merged.append(range) # add the range to list of merged and unmergable ranges
        ranges.remove(range) # remove the range from the list of ranges
    
    # loop through positions and ranges from the list of ranges
    for position, range in enumerate(ranges):

        # at the 0th position, do
        if position == 0:
            current = range # range value becomes current range

        # at other positions, do
        else:

            i = current[0] # start of current range
            j = current[1] # end of current range

            k = range[0] # start of to be compared range
            l = range[1] # end of to be compared ranef
            
            # overalapping and neighbouring intervals are considered
            if i <= k and k <= j+1:
                
                check_result = [min(i, k), max(j, l)] # merged interval takes the minimum and maximum values out of the 2 ranges

            # overalapping and neighbouring intervals are considered - opposite side
            elif k <= i and i <= l+1:

                check_result = [min(i, k), max(j, l)] # merged interval takes the minimum and maximum values out of the 2 ranges

            # otherwise, do
            else:
                check_result = False # merged becomes False

            # when the result of if statements is other than False
            if check_result != False: 

                ranges.append(check_result) # add the merged interval back to the list of ranges

                ranges.remove(current) # remove current
                ranges.remove(range) # remove the one being compared
                break
            
            # otherwise, do
            else:
                
                # when the last position is reached and no merge occurred, do
                if position == len(ranges)-1:

                    merged.append(current) # add the unmergable range among list of porcessed ranges - "merged"
                    ranges.remove(current) # remove the unmergable range from the list of ranges

                continue

# accumulator variable for summing up the amount of fresh IDs
fresh_ingredients = 0      

# loop through processed ranges, get first and last positions of each range in each iteration
for first, last in merged:
    fresh_ingredients += last - first + 1 # add the number of fresh IDs per range; + 1 = closed interval     

# print the results
print("Total number of ingredients to be considered fresh:", fresh_ingredients)     

Total number of ingredients to be considered fresh: 332067203034711


## Answer
332067203034711